akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -469,36 +469,28 @@ object ByteString {
 
     override def drop(n: Int): ByteString =
       if (n <= 0) this
-      else if (n > length) ByteString.empty
+      else if (n >= length) ByteString.empty
       else drop0(n)
 
-    private def drop0(n: Int): ByteString = {
-      var continue = true
-      var fullDrops = 0
-      var remainingToDrop = n
-      do {
-        // impl note: could be optimised a bit by using VectorIterator instead, 
-        //            however then we're forced to call .toVector which halfs performance
-        //            We can work around that, as there's a Scala private method "remainingVector" which is fast, 
-        //            but let's not go into calling private APIs here just yet.
-        val currentLength = bytestrings(fullDrops).length
-        if (remainingToDrop >= currentLength) {
-          fullDrops += 1
-          remainingToDrop -= currentLength
-        } else continue = false
-      } while (remainingToDrop > 0 && continue)
-
-      val remainingByteStrings = bytestrings.drop(fullDrops)
-      if (remainingByteStrings.isEmpty) ByteString.empty
-      else if (remainingToDrop > 0) {
-        val h: ByteString1 = remainingByteStrings.head.drop1(remainingToDrop)
-        val bs = remainingByteStrings.tail
-
-        if (h.isEmpty)
-          if (bs.isEmpty) ByteString.empty
-          else new ByteStrings(bs, length - n)
-        else new ByteStrings(h +: bs, length - n)
-      } else ByteStrings(remainingByteStrings, length - n)
+    private[akka] def drop0(n: Int): ByteString = {
+      // impl note: could be optimised a bit by using VectorIterator instead,
+      //            however then we're forced to call .toVector which halfs performance
+      //            We can work around that, as there's a Scala private method "remainingVector" which is fast,
+      //            but let's not go into calling private APIs here just yet.
+      @tailrec def go(head: Int, restToDrop: Int): (Int, Int) = {
+        val bs = bytestrings(head)
+        if (bs.length > restToDrop) (head, restToDrop)
+        else go(head + 1, restToDrop - bs.length)
+      }
+
+      val (head, restToDrop) = go(0, n)
+
+      if (restToDrop == 0)
+        new ByteStrings(bytestrings.drop(head), length - n)
+      else if (head == bytestrings.length - 1)
+        bytestrings(head).drop(restToDrop)
+      else
+        new ByteStrings(bytestrings(head).drop1(restToDrop) +: bytestrings.drop(head + 1), length - n)
```

Comment:
@ktoso 
hmm, yes it might be inconsistent. It seems like you implemented some related code, say `dropRight1`? Did you name it as is because it belongs to `ByteString1`? or you didn't have a strong reason to name them? :)  If there's no need for us to make a difference between 0 or 1, then we can make them consistent I suppose :)

@johanandren 
Yeah, that'd be better ;)
P.S.
A minor English thing. To me it seems a little bit weird to use `remaining` as a noun here. Most of the cases it's used as an adjective like [Oxford dictionaries](https://en.oxforddictionaries.com/definition/remaining) says? (I know it can be used as a noun though) Anyway I don't have a strong disapproval of it, so let's follow `Scala#Vector`'s naming convention ;)